[Flang][OpenMP] Deprecate Allocate Directive

Stylie777 opened the pull request:
As part of OpenMP 5.2, the allocate directive has been deprecated in favour of the allocators construct for Fortran when an ALLOCATE statement follows the OpenMP allocate directive.

To enable this in flang, a warning has been added informing the user of this. Tests to ensure this behaviour is continued are also included.

See also: #110008

Stylie777 committed:
[Flang][OpenMP] Deprecate Allocate Directive

As part of OpenMP 5.2, the allocate directive has been deprecated
in favour of the allocators construct for Fortran. To enable this in
flang, a warning has been added informing the user of this. Tests to
ensure this behaviour is continued are also included.

See also: #110008

Stylie777 committed:
Address review comments

The following has been fixed:
- Updated comments to be `!` instead of `\\` in test file

Stylie777 commented on flang/test/Lower/OpenMP/Todo/omp-declarative-allocate.f90:
```diff
@@ -5,6 +5,7 @@
   integer :: x, y
 
+  // CHECK: OpenMP directive ALLOCATE has been deprecated, please use ALLOCATORS instead.
   // CHECK: not yet implemented: OpenMPDeclarativeAllocate
```
Updated

Stylie777 committed:
fix failing test

kparzysz commented:
Only the form of ALLOCATE directive that was associated with an ALLOCATE statement was deprecated:
[5.2:625:9-10] "The use of one or more allocate directives with an associated ALLOCATE statement was deprecated."

kparzysz commented:
In terms of AST nodes, it's the `OpenMPExecutableAllocate` form that is deprecated.

Stylie777 committed:
Remove deprecation warning for OpenMPDeclarativeAllocate

As per the spec for OpenMP 5.2, it is only when the allocate
directive is followed by an allocate statement that has been
deprecated. This commit removes the deprecation for the declarative
version of `allocate`, and updates the appropriate tests to
accomodate this.

Stylie777 commented:
Thanks @kparzysz and @raghavendhra for your review.
I have reworked the PR to only add the deprecation warning for executable allocate directives.

kparzysz reviewed (commented):
This looks good, thanks.  Could you add a version check, and only emit the warning when the OpenMP version is >= 52?  You can get the version from SemanticsContext:
`context_.langOptions().OpenMPVersion`.

tblah commented:
> This looks good, thanks. Could you add a version check, and only emit the warning when the OpenMP version is >= 52? You can get the version from SemanticsContext: `context_.langOptions().OpenMPVersion`.

+1 but we decided not to that in this patch because there's no version check for the other deprecation warnings. Would it be okay to fix those in a separate patch?

Stylie777 commented:
Thanks everyone for your reviews.
